### ISSUE TITLE
dck-redux saga/items/actions helpers

### ISSUE THREAD

**Issue by ambroseus** [AgileVisionCompany/dck#57]
it will be good to include in dck-redux our helpful helpers :)
1. Actions/Reducer/Selectors to set/get item's related data, such as for ex. validationStatus and totalPages
2. CRUD action helpers for takeEvery saga effect
3. Process saga helpers

**Comment by ambroseus** [comment]
Actions/Reducer/Selectors to set/get item's related data, such as for ex. validationStatus and totalPages
```javascript
export const ActionTypes = {
  SET_ITEM_DATA: "SET_ITEM_DATA"
};

const actionSetItemData = (itemType, field, data) => ({
  type: ActionTypes.SET_ITEM_DATA,
  itemType,
  field,
  data
});

export const ActionCreators = {
  setItemData: actionSetItemData,
  setValidateStatus: (itemType, validateStatus) =>
    actionSetItemData(itemType, "validateStatus", validateStatus),
  setTotalPages: (itemType, totalPages) =>
    actionSetItemData(itemType, "totalPages", totalPages)
};
```
```javascript
export const Reducer = (state = {}, action) => {
  const itemField = action.field;
  const stateItemData =
    itemField && state[itemField] ? state[itemField] || {} : {};
  switch (action.type) {
    case ActionTypes.SET_ITEM_DATA:
      return {
        ...state,
        [itemField]: {
          ...stateItemData,
          [action.itemType]: action.data
        }
      };
    default:
      return state;
  }
};
```
```javascript
const getFieldData = (state, itemType, field) =>
  (itemType && state.app && state.app[field] && state.app[field][itemType]) ||
  null;
export const appSelectors = {
  getItemData: getFieldData,
  getValidateStatus = (state, itemType) =>
    getFieldData(state, itemType, "validateStatus"),
  getTotalPages = (state, itemType) =>
    getFieldData(state, itemType, "totalPages");

```

**Comment by ambroseus** [comment]
CRUD action helpers for takeEvery saga effect
```javascript
/*
helpers for take* sagas effects

without helper:
  takeEvery(
    action =>
      action.type === DckActionTypes.ITEMS_LOAD &&
      action.itemType === ItemType,
    loadItemsSaga
  )

with helper:
  takeEvery(isActionLoad(ItemType), loadItemsSaga)
*/

const checkActionType = actionType => itemType => action =>
  action.type === actionType && action.itemType === itemType;

const checkActionTypes = actionTypes => itemType => action =>
  actionTypes.includes(action.type) && action.itemType === itemType;

export const isActionLoad = checkActionType(DckActionTypes.ITEMS_LOAD);
export const isActionAdd = checkActionType(DckActionTypes.ITEM_ADD);
export const isActionUpdate = checkActionTypes([
  DckActionTypes.ITEM_SAVE,
  DckActionTypes.ITEMS_SAVE
]);
export const isActionDelete = checkActionType(DckActionTypes.ITEMS_REMOVE);
export const isActionImport = checkActionType(DckActionTypes.ITEMS_IMPORT);
export const isActionSelect = checkActionType(DckActionTypes.ITEM_SELECT);
```

**Comment by ambroseus** [comment]
Process saga helpers
```javascript
import { put, select } from "redux-saga/effects";
import { DckActionCreators } from "dck-redux";
import { ActionCreators } from "../actions";
import { start, reset, stop, fail } from "./helpers";
import { getSessionData } from "../saga/account";
import * as Selectors from "../actions/selectors";

export class Process {
  constructor(processCode, itemType) {
    this.processCode = itemType
      ? `${itemType}_${processCode}`.toUpperCase()
      : processCode;
    this.itemType = itemType || processCode;
  }
  session = getSessionData;
  filters = () => select(Selectors.filters(this.itemType));
  sorting = () => select(Selectors.sorting(this.itemType));

  start = () => put(start(this.processCode));
  reset = () => put(reset(this.processCode));
  stop = description => put(stop(this.processCode, description));
  fail = error => put(fail(this.processCode, error));

  active = id => {
    if (id) return put(DckActionCreators.itemMakeActive(this.itemType, id));
  };

  set = (data, id) => {
    if (!data) return;
    if (id) {
      return put(DckActionCreators.itemSet(this.itemType, id, data));
    } else {
      if (!Array.isArray(data)) data = [data];
      return put(DckActionCreators.itemsSet(this.itemType, data));
    }
  };

  setPages = totalPages => {
    return put(ActionCreators.setTotalPages(this.itemType, totalPages));
  };
}

/* usage example:

function* loadOrdersSaga() {
  const proc = new ProcessLoad(ItemTypes.Order);
  yield proc.start();
  try {
    const session = yield proc.session();
    const filters = yield proc.filters();
    const sorting = yield proc.sorting();

    const response = yield call(
      RestApi.GetOrders,
      session.idToken,
      filters,
      sorting
    );
    yield proc.set(response.data);
    if (response.data.length > 0) yield proc.active(0);
    yield proc.stop();
  } catch (ex) {
    yield proc.fail(ex);
  }
}
*/

export class ProcessLoad extends Process {
  constructor(itemType) {
    super("load", itemType);
  }
}
export class ProcessAdd extends Process {
  constructor(itemType) {
    super("add", itemType);
  }
}

export class ProcessUpdate extends Process {
  constructor(itemType) {
    super("update", itemType);
  }
}

export class ProcessDelete extends Process {
  constructor(itemType) {
    super("remove", itemType);
  }
}

export class ProcessImport extends Process {
  constructor(itemType) {
    super("import", itemType);
  }
  loadImportStatus = options =>
    put(DckActionCreators.itemsLoad(`${this.itemType}ImportStatus`, options));
}

Process.Load = ProcessLoad;
Process.Add = ProcessAdd;
Process.Update = ProcessUpdate;
Process.Delete = ProcessDelete;
Process.Import = ProcessImport;

```

**Comment by ambroseus** [comment]
setItemData/getItemData actions/reducers/selectors have been implemented in dck-redux v4.0
